Add `AccessibleRole::Image` and use it in the `AboutSlint` widget

DataTriny opened the pull request:
<!--
- [x] If the change modifies a visible behavior, it changes the documentation accordingly
- [x] If possible, the change is auto-tested
- [ ] If the changes fixes or close an existing issue, the commit message reference the issue with `Fixes #xxx` or `Closes #xxx`
- [x] If the change is noteworthy, the commit message should contain `ChangeLog: ...`
-->

Introduces a new `AccessibleRole` to represent images and maps it to the appropriate role for each accessibility backend. The role could be set by default but (1) it would probably have to be done in a compiler pass and (2) the `Image` widget is already used extensively to represent icons and other visual elements that really should not be part of the accessibility tree, so it would be annoying to set `accessible-role: none` everywhere.

I took this opportunity to improve the `AboutSlint` widget. I chose "#MadeWithSlint" as the label because I think this is the most important piece of information carried out by the image, but this is your branding so feel free to suggest something else.

ogoffart commented:
We set the default role for Text there. https://github.com/slint-ui/slint/blob/2e4353d3c952b3da7d741b535e49d1e71328f738/internal/compiler/passes/lower_accessibility.rs#L63

But it is true that if in practice, the images won't have a label by default so it might not be so useful to apply the role if no label is set?
I'll let you be the judge of that.

The patch looks good but perhaps the documentation of the image role could be extended a bit to explain that it is not set by default (if we choose not to set it by default) and what kind of image it should apply to in combination to what other accessible properties.

tronical reviewed (approved):
My guts feeling is that it makes sense to let this be opt-in: exposing the image to assistive tech when there is a label, and otherwise not.

(I thought perhaps it would make sense to have a default for the label extracted from image metadata, but that may not be translated or contextually correct)

tronical commented:
Perhaps via grabbing the pixels from the framebuffer? The bounds of the image/node within the window are known.

DataTriny commented:
> Perhaps via grabbing the pixels from the framebuffer? The bounds of the image/node within the window are known.

Yes I think that's exactly how it works. But for that the assistive technology have to know that there is a node on which it can perform image recognition, hence this new role.

I'll do the documentation tomorrow.

DataTriny commented:
I'm not entirely sure whether images should be part of the accessibility tree by default or not, and I wouldn't want to make a bad choice for Slint. I'm going to reach out to more knowledgeable people to discuss this matter and I'll update you here.

DataTriny commented:
I'm setting the accessible role on all images and I've hidden icons on widgets.

There should probably be a test somewhere to check that the role is set by default, but I'm not sure where would it live?

ogoffart commented on docs/astro/src/content/docs/reference/elements/image.mdx:
```diff
@@ -251,3 +251,29 @@ Use the <Link type="ImageType" label="`@image-url` macro" /> to specify the imag
 Properties in source image coordinates that define the region of the source image that is rendered.
 By default the entire source image is visible:
+
+## Accessibility
```
I'd mention somewhere in this paragraph that Image have a default accessible-role

ogoffart reviewed (approved):
Thanks a lot.
This looks good.

Just an idea, but you'll tell me if it is a good or a bad one: what if the role was only set if the image had a `accessible-label` property set?
OR maybe that's too magical and confusing and we still want the screen reader to know of all image by default even if they don't have a label.

> There should probably be a test somewhere to check that the role is set by default, but I'm not sure where would it live?

Maybe in tests/cases/elements/image.slint ? Add something like `img1.accessible-role == AccessibleRole.image`  in the `test` property.

DataTriny commented:
I've updated the first commit to keep a clean git history. I've added Rust and C++ tests in ` tests/cases/elements/image.slint` and I have updated the documentation to mention that images get a default accessible role before showing how to change this if desired.

> what if the role was only set if the image had a accessible-label property set?

This would break a currently defined rule which prevents setting any accessibility related property unless `accessible-role` is also set, so this would be confusing I guess. This would also make it impossible to have an image without a label. Even though this is something we want to ignore, we still may want to let the user do that?

DataTriny commented:
Can't get the C++ test to work apparently. Well it's not my thing, I guess one of you can spot the obvious mistake in there.

ogoffart commented on tests/cases/elements/image.slint:
```diff
@@ -33,15 +33,26 @@ const TestCase &instance = *handle;
+auto img_search = slint::testing::ElementHandle::find_by_element_id(handle, "TestCase::img");
+assert_eq(img_search.size(), 1);
+auto img = img_search[0];
+assert_eq(*img.accessible_role(), slint::testing::AccessibleRole::Image);
```
```suggestion
assert(*img.accessible_role() == slint::testing::AccessibleRole::Image);
```

the error is that slint::testing::AccessibleRole doesn't implement the stream operator (equivalent to Debug trait in rust)

ogoffart commented:
Thanks!
